syntax = "proto3";
package Raft;

service RaftService {
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
}

message RequestVoteRequest {
    uint64 term = 1;
    uint64 candidateId = 2;
    uint64 lastLogIndex = 3;
    uint64 lastLogTerm = 4;
}

message RequestVoteResponse {
    uint64 term = 1;
    bool voteGranted = 2;
}

message LogEntry {
    uint64 term = 1;
    uint64 index = 2;
    string operation = 3;  // "PUT" or "DELETE"
    string key = 4;
    string value = 5;
}

message AppendEntriesRequest {
    uint64 term = 1;
    uint64 leaderId = 2;
    uint64 prevLogIndex = 3;
    uint64 prevLogTerm = 4;
    repeated LogEntry entries = 5;
    uint64 leaderCommit = 6;
}

message AppendEntriesResponse {
    uint64 term = 1;
    bool success = 2;
}
